### ISSUE TITLE
Wrong MHQ time

### ISSUE THREAD

**Issue by mirabilos** [CuppaZee/CuppaZee#14]
Second timezone bug, just reverse as in the normal Munzee äpp (ZeeOps timeout, they’re not fixing it, they told me…): both the CuppaZee äpp and https://max.cuppazee.app/ show 23:01 MHQ when the MHQ day has just changed.

You *cannot* derive MHQ time from local time by a fixed offset. You have to do the calculations in the MHQ timezone (I believe `America/Chicago` in tzdata). Only that will _also_ keep up with changing transition rules, too…

**Comment by sohcah** [comment]
I am aware that CuppaZee's timezone handling is not ideal, however I'd be interested to hear what timezone you are in so I can try to replicate the issue you are having as it works fine for me here.

I would have preferred for me to use tzdata via the ECMAScript `Intl` API, however when CuppaZee Max V2.2 was released, React Native with Hermes did not have support for Intl.

As such, CuppaZee is a setup to get the current UTC time and offset using the data for America/Chicago manually embedded, see https://github.com/CuppaZee/CuppaZee/blob/main/packages/utils/src/dayjsmhq.ts#L12-L730, which should give correct values in all cases.

Unfortunately, Hermes without Intl isn't very reliable at getting your device's current timezone offset and as such doesn't handle nicely getting the current UTC time or converting timestamps between timezones.

With CuppaZee V3.0, I'll be upgrading to a new version of Hermes which finally supports Intl, which should eliminate most of these issues, however V3 is still probably a few months away due to me being busy with College.

It should also be noted that the Time displayed in CuppaZee's header doesn't re-render consistently, so can in many cases display a time from when you first opened the app many hours or even days ago. 

Really, the optimal solution would be to use the ECMAScript Temporal API when that releases, however I believe there are a few major blockers on that before it reaches Stage 4 and can be implemented by browsers as well as React Native/Hermes and Node.JS, so that's also a long way off.

**Comment by mirabilos** [comment]
Huh. This is rather hard to debug on CuppaZee itself for me.

But when I go to https://day.js.org/ and type `dayjs().utcOffset()` into the F12 console, I get 120, which… probably is the expected value, if thats minutes. As a Unix developer I would be expecting seconds. TZ is Europe/Berlin right now, and max.cuppazee.app shows the date off by one hour still. (We just had a DST transition this weekend, and America’s has been some time ago.)

**Comment by sohcah** [comment]
Yes, in JS typically we work with UTC offsets in minutes... and everything else in Milliseconds.

I've tried setting my timezone to Europe/Berlin on my Mac (with CuppaZee Max in MS Edge Browser), my Windows machine (with CuppaZee Max in Vivaldi Browser) and on my iPhone (with the CuppaZee Max iOS app), and I'm unable to reproduce this.

Would you be able to confirm for me what Browser and OS you are using so I can attempt to reproduce the issue on that platform?

**Comment by mirabilos** [comment]
Sam Hindess dixit:

>Would you be able to confirm for me what Browser and OS you are using

Debian bullseye/amd64, firefox-esr 91.7.0esr-1~deb11u1

Also, iOS 15.3.1 and Äpp version… where do I see that?

>so I can attempt to reproduce the issue on that platform?

But, do you know what… it seems to have self-healed on both.
Do you perhaps cache this value? (If so… safest is to invalidate
every full *and* half (for Indian timezones) hour as obtaining it
shouldn’t be _too_ costly.)

bye,
//mirabilos
-- 
„Cool, /usr/share/doc/mksh/examples/uhr.gz ist ja ein Grund,
mksh auf jedem System zu installieren.“
	-- XTaran auf der OpenRheinRuhr, ganz begeistert
(EN: “[…]uhr.gz is a reason to install mksh on every system.”)
